Add "MayRunAs" value among other GroupStrategies

stlaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Adds "MayRunAs" value among other group strategies. This strategy
allows to define a certain range of GIDs for FSGroupStrategy and
SupplementalGroupStrategy in a PSP.

This new strategy works similarly to the "MustRunAs" one, except that
when no GID is specified in a pod/container security context then no
GID is generated for the respective containers.

**Which issue(s) this PR fixes**
Resolves #56173

**Release note**:
```release-note
PodSecurityPolicy objects now support a `MayRunAs` rule for `fsGroup` and `supplementalGroups` options. This allows specifying ranges of allowed GIDs for pods/containers without forcing a default GID the way `MustRunAs` does. This means that a container to which such a policy applies to won't use any fsGroup/supplementalGroup GID if not explicitly specified, yet a specified GID must still fall in the GID range according to the policy.
```

stlaz committed:
Add "MayRunAs" value among other GroupStrategies

Adds "MayRunAs" value among other group strategies. This strategy
allows to define a certain range of GIDs for FSGroupStrategy and
SupplementalGroupStrategy in a PSP.

This new strategy works similarly to the "MustRunAs" one, except that
when no GID is specified in a pod/container security context then no
GID is generated for the respective containers.

Resolves #56173

php-coder commented:
/sig auth
/ok-to-test

PTAL @kubernetes/sig-auth-api-reviews 

CC @simo5

php-coder commented on pkg/security/podsecuritypolicy/group/mayrunas.go:
```diff
@@ -0,0 +1,80 @@
+
+var _ GroupStrategy = &mayRunAs{}
+
+// NewRunAsAny provides a new RunAsAny strategy.
```
Please, update the comment.

php-coder commented on pkg/security/podsecuritypolicy/group/mayrunas.go:
```diff
@@ -0,0 +1,80 @@
+// NewRunAsAny provides a new RunAsAny strategy.
+func NewMayRunAs(ranges []policy.IDRange, field string) (GroupStrategy, error) {
+	if len(ranges) == 0 {
+		return nil, fmt.Errorf("ranges must be supplied for MustRunAs")
```
s/MustRunAs/MayRunAs/

php-coder commented on pkg/apis/policy/types.go:
```diff
@@ -389,6 +392,9 @@ type SupplementalGroupsStrategyOptions struct {
 type SupplementalGroupsStrategyType string
 
 const (
+	// SupplementalGroupsStrategyMayRunAs means that container does not need to run as a particular gid.
```
s/to run as a particular gid/to run with particular gids/ ?

stlaz replied:
It was but a mere copy-paste but "with" sounds definitely better

php-coder commented on pkg/apis/policy/types.go:
```diff
@@ -389,6 +392,9 @@ type SupplementalGroupsStrategyOptions struct {
 
 const (
+	// SupplementalGroupsStrategyMayRunAs means that container does not need to run as a particular gid.
+	// However, when it is applied, it has to fall in the defined range.
```
s/when it is applied, it has to/when they are specified, they have to/ ?

php-coder commented:
@stlaz One more place to update:

https://github.com/kubernetes/kubernetes/blob/3abba25160590921fec61236ba012a8bbd757d6c/pkg/apis/policy/fuzzer/fuzzer.go#L49-L58

php-coder commented:
@kubernetes/sig-auth-api-reviews Do we need to update PSP in `extensions` API group?

php-coder commented:
/assign @tallclair

stlaz commented:
Updated the patch to fix `verify-basel` and failure and to address the PR comments.

stlaz commented:
This change introduces backward incompatibility since the "MayRunAs" value to the groups is unknown to previous API versions. If we convert this to "MustRunAs" internally, we wouldn't be able to later get the information that it's actually "MayRunAs".
What would be the preferred way to handle such an issue? Is the rollback desirable here? Does @tallclair have an input on this?

krmayankk commented:
@tallclair @php-coder i believe the MayRunAs would also be needed by RunAsGroup field. I will include that in the PSP changes for RunAsGroup in my next PR.

php-coder commented:
PTAL @tallclair

tallclair commented on pkg/security/podsecuritypolicy/group/mayrunas.go:
```diff
@@ -0,0 +1,80 @@
+func (s *mayRunAs) Validate(_ *api.Pod, groups []int64) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, group := range groups {
```
dedupe this code with `mustRunAs.Validate` (extract a shared helper)

tallclair commented on pkg/security/podsecuritypolicy/group/mayrunas_test.go:
```diff
@@ -0,0 +1,161 @@
+	}
+}
+
+func TestMayRunAsGenerate(t *testing.T) {
```
This method is constant, don't bother testing it here. Instead, add a couple cases under https://github.com/kubernetes/kubernetes/blob/master/pkg/security/podsecuritypolicy/provider_test.go to ensure that the nil return works as expected.

tallclair replied:
Ping.

tallclair reviewed (commented):
> This change introduces backward incompatibility since the "MayRunAs" value to the groups is unknown to previous API versions.

Backwards incompatibility means that a resource that was valid in a previous version is no longer valid (or behaves differently), which isn't the case here. What you're refering to is _forwards compatibility_, and that's not something Kubernetes provides, so this is a non-issue.

stlaz commented:
Thank you @tallclair for your comments, my apologies that it took me so long to react in the PR accordingly.
I added a bunch of tests to `podsecuritypolicy/provider_test.go` and removed most of `mayrunas_test.go::TestMayRunAsGenerate` (kept only one to test the behavior is correct).

tallclair commented on pkg/security/podsecuritypolicy/group/helpers.go:
```diff
@@ -0,0 +1,39 @@
+
+func CheckGroupsInRanges(ranges []policy.IDRange, groups []int64, fieldName string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	errFmt := "group %d must be in the ranges: %v"
```
nit: const

tallclair commented on pkg/security/podsecuritypolicy/group/helpers.go:
```diff
@@ -0,0 +1,39 @@
+	psputil "k8s.io/kubernetes/pkg/security/podsecuritypolicy/util"
+)
+
+func CheckGroupsInRanges(ranges []policy.IDRange, groups []int64, fieldName string) field.ErrorList {
```
nit: prefer `ValidateGroupsInRanges`

tallclair reviewed (commented):
A few nits, and please fix the test. Everything else lgtm.

stlaz commented:
Sorry about the test failures, was in a hurry and did not get to address them. Should hopefully be fixed now. I also removed the rest of `TestMayRunAsGenerate()`.

stlaz commented on pkg/security/podsecuritypolicy/group/helpers.go:
```diff
@@ -0,0 +1,39 @@
+
+	for _, group := range groups {
+		for _, rng := range ranges {
+			if !psputil.GroupFallsInRange(group, rng) {
```
Am a bit of a newbie to Go - how exactly do you want me to inline this?

stlaz replied:
Ah, I see, I did not check the contents of the function. I'd say let's leave it here for now as is if you don't mind.

tallclair commented:
/lgtm

stlaz commented:
Rebased on the current master after the field path rework.

stlaz commented:
/test pull-kubernetes-e2e-gce

tallclair commented:
Please either squash the commits, or split all the autogen files into a separate commit.

stlaz commented:
Did not notice the autogenerated file to have appeared there, should be fixed in the latest patchset.

stlaz commented:
/test pull-kubernetes-e2e-kops-aws

stlaz commented:
/retest

stlaz commented:
/retest

stlaz commented:
/retest

stlaz commented:
It would seem that we're continuously unable to get enough resources for the last test to pass, is that so?
```
error running task "VPC/e2e-97946-dba53.test-cncf-aws.k8s.io" (9m56s remaining to succeed): error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.
```

stlaz commented:
/retest
Let's give it a try again.

tallclair commented:
/lgtm
/milestone v1.12
/status approved-for-milestone

stlaz commented:
/retest
/kind feature
@erictune this PR still needs approval. Please see if it's approvable.
Also, I suspect that you people decide on the priority of the PR :slightly_smiling_face:

stlaz commented:
/priority important-longterm
This one seems to match, feel free to change.

stlaz commented:
@liggitt can you approve or get someone to approve if good enough?
@erictune does not seem to be available and the bot will probably just re-run the test every 96 hours for no reason unless pushed/closed.

stlaz commented:
/test pull-kubernetes-e2e-kops-aws

stlaz commented:
@erictune PTAL

stlaz commented:
/sig apps-api-approvers

php-coder commented:
PTAL @kubernetes/sig-apps-pr-reviews

stlaz commented:
/test pull-kubernetes-e2e-kops-awS

stlaz commented:
@erictune or anyone from the sig-apps-pr-reviews group, please see about approving this PR.

tallclair commented:
/unassign @erictune 
/assign @liggitt

stlaz commented:
Thank you for the assignee swap :)
@liggitt, could you please have a look and see if this PR is good for approval?

jimangel commented:
@stlaz do you have a docs PR against the 1.12 branch?

stlaz commented:
I do now: https://github.com/kubernetes/website/pull/9888

stlaz commented:
Thanks for the docs review :) Only needs an approval now.
@liggitt would you see if you can approve this PR, please?

stlaz commented:
@liggitt PTAL

liggitt commented on pkg/security/podsecuritypolicy/group/helpers.go:
```diff
@@ -0,0 +1,40 @@
+	allErrs := field.ErrorList{}
+
+	for _, group := range groups {
+		for _, rng := range ranges {
```
Is this correct? I thought the group would be required to fall within one of the ranges (like isGroupValid checks), not required to fall within all of them

stlaz replied:
Hm, you're right, I just blindly copy-pasted the code from `MustRunAs` originally, but it would be my impression that falling in just one of the ranges should be enough. I'll definitely add a test for that. It's weird that noone noticed this up until now.

stlaz replied:
Actually, never mind, it seems the original code was correct and I am missing a break here.

liggitt commented on pkg/security/podsecuritypolicy/group/mayrunas.go:
```diff
@@ -0,0 +1,69 @@
+// Groups are passed in here to allow this strategy to support multiple group fields (fsgroup and
+// supplemental groups).
+func (s *mayRunAs) Validate(fldPath *field.Path, _ *api.Pod, groups []int64) field.ErrorList {
+	return ValidateGroupsInRanges(fldPath, s.ranges, groups)
```
See question about whether the "match all ranges" logic in ValidateGroupsInRanges is correct. If not, please add a unit test with two disjoint ranges and a valid group that falls within one of the ranges

liggitt commented on pkg/security/podsecuritypolicy/group/mayrunas_test.go:
```diff
@@ -0,0 +1,115 @@
+		},
+		"not in range": {
+			groups: []int64{5},
+			ranges: []policy.IDRange{
```
Need a test with multiple ranges with a valid group

liggitt commented on pkg/security/podsecuritypolicy/group/mustrunas.go:
```diff
@@ -66,12 +66,7 @@ func (s *mustRunAs) Validate(fldPath *field.Path, _ *api.Pod, groups []int64) fi
 	}
 
-	for _, group := range groups {
-		if !s.isGroupValid(group) {
```
I think this switch is incorrect, and reveals a gap in unit testing

liggitt commented on pkg/security/podsecuritypolicy/provider_test.go:
```diff
@@ -604,23 +631,37 @@ func TestValidatePodSecurityContextSuccess(t *testing.T) {
+	supGroupMayPSP.Spec.SupplementalGroups = policy.SupplementalGroupsStrategyOptions{
+		Rule: policy.SupplementalGroupsStrategyMayRunAs,
+		Ranges: []policy.IDRange{
+			{Min: 1, Max: 5},
```
Make a positive test with multiple ranges

liggitt commented:
/hold
for resolution of requiring a group to match all ranges. LGTM otherwise

stlaz commented:
/retest

stlaz commented:
The "fall in all groups" issue should now be resolved, a positive multi-ranges test was added on top of the mayrunas tests.

liggitt commented:
thanks. both tallclair and I are out of the office at the moment, but will pick this up when master reopens for 1.13.

stlaz commented:
Thank you. I'm not sure exactly why the verify test does not pass since `make verify` seems to be running fine on my system. The only guess I have was my upgrade of golang from 1.10 to 1.11 :thinking:

stlaz commented:
/retest

tallclair commented on pkg/security/podsecuritypolicy/group/mayrunas_test.go:
```diff
@@ -0,0 +1,122 @@
+				{Min: 1, Max: 4},
+				{Min: 10, Max: 15},
+			},
+		},
```
please add cases with multiple groups.

tallclair reviewed (commented):
A couple nits, but LGTM. Please squash.

stlaz commented:
Comments addressed + squashed.

tallclair commented:
/lgtm
/hold cancel

yue9944882 commented:
hi folks, why not syncing these new internal definitions to the external repo? i'm afraid currently we can't specify the new strategy w/ client-go. also the "MayRunAs" will be an undefined value when deserialized into external types.
